diff --git a/cuda_bindings/build_hooks.py b/cuda_bindings/build_hooks.py
index e83485fc4..3d3849ad6 100644
--- a/cuda_bindings/build_hooks.py
+++ b/cuda_bindings/build_hooks.py
@@ -415,7 +415,7 @@ def _build_cuda_bindings(strip=False):
     _extensions = cythonize(
         extensions,
         nthreads=nthreads,
-        build_dir="build/cython",
+        build_dir="." if compile_for_coverage else "build/cython",
         compiler_directives=cython_directives,
         **extra_cythonize_kwargs,
     )
diff --git a/cuda_bindings/pyproject.toml b/cuda_bindings/pyproject.toml
index 558fe1aa3..5c8d0f30a 100644
--- a/cuda_bindings/pyproject.toml
+++ b/cuda_bindings/pyproject.toml
@@ -62,7 +62,6 @@ include = ["cuda*"]
 include-package-data = false
 
 [tool.setuptools.package-data]
-"*" = ["*.pxd", "*.pxi"]
 "cuda.bindings._bindings" = ["*.h"]
 "cuda.bindings._lib" = ["*.h"]
 
diff --git a/cuda_bindings/setup.py b/cuda_bindings/setup.py
index adbf843fd..dc460f331 100644
--- a/cuda_bindings/setup.py
+++ b/cuda_bindings/setup.py
@@ -46,10 +46,16 @@ class build_ext(_build_ext):
         super().build_extensions()
 
 
+coverage_mode = os.environ.get("CUDA_PYTHON_COVERAGE", "0") != "0"
+package_data = {"": ["*.pxd", "*.pxi"]}
+if coverage_mode:
+    package_data[""] += ["*.pyx", "*.cpp"]
+
 setup(
     ext_modules=build_hooks._extensions,
     cmdclass={
         "build_ext": build_ext,
     },
+    package_data=package_data,
     zip_safe=False,
 )
diff --git a/cuda_core/build_hooks.py b/cuda_core/build_hooks.py
index 5d615ad00..010133219 100644
--- a/cuda_core/build_hooks.py
+++ b/cuda_core/build_hooks.py
@@ -163,7 +163,7 @@ def _build_cuda_core():
         ext_modules,
         verbose=True,
         language_level=3,
-        build_dir="build/cython",
+        build_dir="." if COMPILE_FOR_COVERAGE else "build/cython",
         nthreads=nthreads,
         compiler_directives=compiler_directives,
         compile_time_env=compile_time_env,
diff --git a/cuda_core/pyproject.toml b/cuda_core/pyproject.toml
index b08f43503..3d68e0596 100644
--- a/cuda_core/pyproject.toml
+++ b/cuda_core/pyproject.toml
@@ -77,7 +77,6 @@ include = ["cuda.core*"]
 include-package-data = false
 
 [tool.setuptools.package-data]
-"*" = ["*.pxd"]
 "cuda.core._include" = ["*.h", "*.hpp"]
 "cuda.core._cpp" = ["*.hpp"]
 
diff --git a/cuda_core/setup.py b/cuda_core/setup.py
index 4a501edc1..ad02428d7 100644
--- a/cuda_core/setup.py
+++ b/cuda_core/setup.py
@@ -17,10 +17,16 @@ class build_ext(_build_ext):
         super().build_extensions()
 
 
+coverage_mode = os.environ.get("CUDA_PYTHON_COVERAGE", "0") != "0"
+package_data = {"*": ["*.pxd"]}
+if coverage_mode:
+    package_data["*"] += ["*.pxi", "*.pyx", "*.cpp"]
+
 setup(
     ext_modules=build_hooks._extensions,
     cmdclass={
         "build_ext": build_ext,
     },
+    package_data=package_data,
     zip_safe=False,
 )
